nexus/inventory/src/builder.rs

```diff
@@ -0,0 +1,278 @@
+                            slot_a_sha3_256_digest,
+                            slot_b_sha3_256_digest,
+                        }
+                    });
+            }
+            gateway_client::types::RotState::CommunicationFailed {
+                message,
+            } => {
+                self.found_error(anyhow!(
+                    "MGS {:?}: reading RoT state for {:?}: {}",
+                    source,
+                    baseboard,
+                    message
+                ));
+            }
+        }
+
+        Some(baseboard)
+    }
+
+    /// Returns true if we already found the caboose for `which` for baseboard
+    /// `baseboard`
+    ///
+    /// This is used to avoid requesting it multiple times (from multiple MGS
+    /// instances).
+    pub fn sp_found_caboose_already(
+        &self,
+        baseboard: &BaseboardId,
+        which: CabooseWhich,
+    ) -> bool {
+        self.cabooses_found
+            .get(&which)
+            .map(|map| map.contains_key(baseboard))
+            .unwrap_or(false)
+    }
+
+    /// Record the given caboose information found for the given baseboard
+    ///
+    /// The baseboard must previously have been reported using
+    /// `found_sp_state()`.
+    ///
+    /// `source` is an arbitrary string for debugging that describes the MGS
+    /// that reported this data (generally a URL string).
+    pub fn found_sp_caboose(
+        &mut self,
+        baseboard: &BaseboardId,
+        which: CabooseWhich,
+        source: &str,
+        caboose: SpComponentCaboose,
+    ) -> Result<(), anyhow::Error> {
+        // Normalize the caboose contents: i.e., if we've seen this exact caboose
+        // contents before, use the same record from before.  Otherwise, make a
+        // new one.
+        let sw_caboose =
+            Self::enum_item(&mut self.cabooses, Caboose::from(caboose));
+        let (baseboard, _) =
+            self.sps.get_key_value(baseboard).ok_or_else(|| {
+                anyhow!(
+                    "reporting caboose for unknown baseboard: {:?} ({:?})",
+                    baseboard,
+                    sw_caboose
+                )
+            })?;
+        let by_id =
+            self.cabooses_found.entry(which).or_insert_with(|| BTreeMap::new());
+        if let Some(previous) = by_id.insert(
+            baseboard.clone(),
+            CabooseFound {
+                time_collected: Utc::now(),
+                source: source.to_owned(),
+                caboose: sw_caboose.clone(),
+            },
+        ) {
+            let error = if *previous.caboose == *sw_caboose {
+                anyhow!("reported multiple times (same value)",)
+            } else {
+                anyhow!(
+                    "reported caboose multiple times (previously {:?}, \
+                    now {:?})",
+                    previous,
+                    sw_caboose
+                )
+            };
+            Err(error.context(format!(
```

Comment:
I don't understand a sequence by which we could see this error message - what is this guarding against?